src/tiled/wangtemplateview.cpp

```diff
@@ -0,0 +1,403 @@
+            p.clear();
+            //right
+            p.append(QPoint(0, 0));
+            p.append(centeredRect.bottomRight());
+            p.append(centeredRect.topRight());
+
+            painter->setBrush(getColor(wangId.edgeColor(1), edgeCount));
+            painter->drawPolygon(p);
+            p.clear();
+            //bottom
+            p.append(QPoint(0, 0));
+            p.append(centeredRect.bottomRight());
+            p.append(centeredRect.bottomLeft());
+
+            painter->setBrush(getColor(wangId.edgeColor(2), edgeCount));
+            painter->drawPolygon(p);
+            p.clear();
+            //left
+            p.append(QPoint(0, 0));
+            p.append(centeredRect.topLeft());
+            p.append(centeredRect.bottomLeft());
+
+            painter->setBrush(getColor(wangId.edgeColor(3), edgeCount));
+            painter->drawPolygon(p);
+        }
+    } else if (cornerCount <= 1) {
+        painter->drawRect(centeredRect);
+    }
+
+    painter->restore();
+}
+
+void WangTemplateDelegate::paint(QPainter *painter,
+                                 const QStyleOptionViewItem &option,
+                                 const QModelIndex &index) const
+{
+    const WangTemplateModel *model = static_cast<const WangTemplateModel*>(index.model());
+    const WangId wangId = model->wangIdAt(index);
+    if(wangId == 0)
+        return;
+
+    if (WangSet *wangSet = mWangTemplateView->wangSet())
+        paintTemplateTile(painter, wangId,
+                      wangSet->edgeColors(),
+                      wangSet->cornerColors(),
+                      option.rect);
+    else
+        paintTemplateTile(painter, wangId, 1, 1, option.rect);
+
+    //Highlight currently selected tile.
+    if (mWangTemplateView->currentIndex() == index) {
+        QColor high = option.palette.highlight().color();
+        painter->setBrush(QColor(high.red(), high.green(), high.blue(), 100));
+        painter->setPen(Qt::NoPen);
+        painter->drawRect(option.rect);
+    }
+
+    //Shade tile if used already
+    if (mWangTemplateView->wangIdIsUsed(wangId)) {
+        painter->setBrush(QColor(0,0,0,100));
+        painter->setPen(Qt::NoPen);
+        painter->drawRect(option.rect.adjusted(2,2,-2,-2));
+    }
+}
+
+QSize WangTemplateDelegate::sizeHint(const QStyleOptionViewItem &option, const QModelIndex &index) const
+{
+    Q_UNUSED(option);
+    Q_UNUSED(index);
+
+    //
+    return QSize(32 * mWangTemplateView->scale(),
+                 32 * mWangTemplateView->scale());
+}
+
+} // anonymous namespace
+
+WangTemplateView::WangTemplateView(QWidget *parent)
+    : QListView(parent)
+    , mZoomable(new Zoomable(this))
+{
+    setWrapping(true);
+    setFlow(QListView::LeftToRight);
+    setResizeMode(QListView::Adjust);
+    setSelectionMode(QAbstractItemView::SingleSelection);
+    setVerticalScrollMode(QAbstractItemView::ScrollPerPixel);
+    setItemDelegate(new WangTemplateDelegate(this, this));
+    setTabKeyNavigation(true);
```

Comment:
Please add:
```c++
setUniformItemSizes(true);
```
I didn't notice an actual performance difference, but the view may be able to avoid getting the size hint very often.

You can remove the `setTabKeyNavigation(true);`. That is the default, but actually it won't work anyway because Tab is used to show/hide the all menu bars and dock widgets. Alternatively it could be set to `false` explicitly.